File: docs/widgets/row/widget.js
// Row container: render each child anywidget side by side.
//
// In a static export, args.host.renderChild(ref, el) mounts a referenced child
// anywidget into our own DOM (kernel-free) and returns its cleanup. The child
// refs arrive as the synced `children` trait (an array of "IPY_MODEL_<id>").

function render({ model, el, host }) {
    el.innerHTML = "";

    const row = document.createElement("div");
    row.className = "myst-row";
    row.style.display = "flex";
    row.style.gap = model.get("gap") || "16px";
    row.style.alignItems = "flex-start";
    row.style.flexWrap = "wrap";
    el.appendChild(row);

    const refs = model.get("children") || [];
    const cleanups = [];

    (async () => {
        for (const ref of refs) {
            const cell = document.createElement("div");
            cell.className = "myst-row__cell";
            row.appendChild(cell);
            if (host && typeof host.renderChild === "function") {
                try {
                    cleanups.push(await host.renderChild(ref, cell));
                } catch (err) {
                    cell.textContent = "[child failed: " + (err && err.message) + "]";
                }
            } else {
                cell.textContent = "[no static host: live kernel path not wired in this demo]";
            }
        }
    })();

    return () => {
        for (const dispose of cleanups) {
            try {
                dispose();
            } catch (err) {
                /* ignore */
            }
        }
    };
}

export default { render };

File: docs/widgets/row/widget.py
"""
A minimal layout container anywidget that arranges child anywidgets side by side.

It exists to demonstrate the static-export container-renderer contract: the
container declares its child-ref traits via `_myst_child_traits`, and its JS
mounts each child with `host.renderChild(ref, el)` so the children render
(still linked) with no kernel attached.
"""
import pathlib

import anywidget
import traitlets
from ipywidgets import Widget, widget_serialization


class Row(anywidget.AnyWidget):
    """Lay out child widgets in a horizontal flex row."""

    _esm = pathlib.Path(__file__).parent / "widget.js"
    _css = pathlib.Path(__file__).parent / "style.css"

    # The children to render side by side.
    children = traitlets.List(trait=traitlets.Instance(Widget)).tag(
        sync=True, **widget_serialization
    )

    # The contract marker the static-export plugin reads: which traits hold the
    # child anywidget refs the container will render itself.
    _myst_child_traits = traitlets.List(["children"]).tag(sync=True)

    gap = traitlets.Unicode("16px").tag(sync=True)

    def __init__(self, children=(), **kwargs):
        super().__init__(children=list(children), **kwargs)
